server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -323,14 +338,7 @@ Index[] concreteIndices(Context context, String... indexExpressions) {
         final Map<String, IndexAbstraction> indicesLookup = context.getState().metadata().getIndicesLookup();
         for (String expression : expressions) {
             final IndexAbstraction indexAbstraction = indicesLookup.get(expression);
-            if (indexAbstraction == null) {
-                continue;
-            } else if (indexAbstraction.getType() == Type.ALIAS && context.getOptions().ignoreAliases()) {
-                continue;
-            } else if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
-                continue;
-            }
```

Comment:
**This is an important change, and I would like to call your attention to it.**

In principle, this moves the "ignoring of unavailable names" logic into the expression resolver inner method, on line 320 above, and away from the concrete indices method. The expression resolver already handles unavailable names when the ignore unavailable option is `false` (and throws an exception) so this improves consistency and groups the handling of explicit and/or unvailable names inside request expressions in a single place in the code, see `ExplicitResourceNameFilter`.

But this requires verification that it is not a problem that the `IndexNameExpressionResolver#resolveExpressions` method will now **not** return (i.e. silently discard) unavailable resources.
I believe this is safe to do.
The 4 other places where this method is called (apart from `concreteIndices` one here) are:
* `IndexNameExpressionResolver#dataStreamNames` this already ignores unavailable names
* `IndexNameExpressionResolver#resolveWriteIndexAbstraction` uses request options that trigger the "not found' exception, so the changed behavior, which only kicks in when said option is not set, doesn't impact it
* the caller `IndexNameExpressionResolver#resolveExpressions` on line 626, is used for computing alias filters in a couple of places. Given the options, it is possible that the change makes it that missing indices are not returned anymore, but missing indices are irrelevant when computing filtering aliases. It shouldn't impact the callers.
* the  usage in `IndexNameExpressionResolver#resolveSearchRouting` uses the resolved expression in computing the search routing per concrete index given routings for the aliases in the expression that refer to the concrete index. It is possible that the change will not return missing indices, but those are again irrelevant in computing the search routing for existing concrete indices, so the change should not impact callers.

Reply:
I think one difference here is that "ignoring of unavailable" was previously done regardless of the context flag, but it seems we now do it only if `context.getOptions().ignoreUnavailable()`. I'm I understanding this correctly? Wouldn't this change behaviour? Perhaps we don't care.

Reply:
> that "ignoring of unavailable" was previously done regardless of the context flag

It's hard to see in the original code, but this is not true.
The unconditional removal of unavailable names coded here is actually going to run only when `ignore_unavailable` is `true`, because if it is `false` it throw the "not found" exception here: https://github.com/elastic/elasticsearch/blob/f1bfd259f2d502bce924b4a8053f7462d621932c/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1222 .
So the if `ignore_unavailable` is `true` is redundant, there can't be "unavailable" names to remove otherwise.

Reply:
To be clear, in other words, in the original code that removed "unavailable" resource names there was no `if (context.getOptions().ignoreUnavailable() == true` guard check, but that was not an issue because that check is redundant. `ignore_unavailable` must be `true` when there are "unavailable" resource names to remove, otherwise earlier processing would throw the "not found" exception.